src/cephadm/cephadm

```diff
@@ -5059,24 +5130,77 @@ def command_ceph_volume(ctx):
         tmp_config = write_tmp(config, uid, gid)
         mounts[tmp_config.name] = '/etc/ceph/ceph.conf:z'
 
+    # Ceph-volume uses the bootstrap-osd key in order to do its operations.
+    # This function retrieves the keyring so it can be provided.
+
+    def get_bootstrap_osd_keyring() -> Optional[str]:
+        if not ctx.keyring and os.path.exists(SHELL_DEFAULT_KEYRING):
+            ctx.keyring = SHELL_DEFAULT_KEYRING
+        (config, keyring) = get_config_and_keyring(ctx)
+
+        mounts = get_container_mounts(ctx, ctx.fsid, 'osd', None,
+                                      no_config=True if ctx.config else False)
+        if ctx.config:
+            mounts[pathify(ctx.config)] = '/etc/ceph/ceph.conf:z'
+        if ctx.keyring:
+            mounts[pathify(ctx.keyring)] = '/etc/ceph/ceph.keyring:z'
+        c = CephContainer(
+            ctx,
+            image=ctx.image,
+            entrypoint='/usr/bin/ceph',
+            args='auth get client.bootstrap-osd'.split(),
+            volume_mounts=mounts,
+        )
+        out, err, code = call_throws(ctx, c.run_cmd())
+        if not code:
+            return out
+        else:
+            return None
+
+    if not keyring:
+        keyring = get_bootstrap_osd_keyring()
+
     if keyring:
         # tmp keyring file
         tmp_keyring = write_tmp(keyring, uid, gid)
         mounts[tmp_keyring.name] = '/var/lib/ceph/bootstrap-osd/ceph.keyring:z'
 
-    c = CephContainer(
-        ctx,
-        image=ctx.image,
-        entrypoint='/usr/sbin/ceph-volume',
-        envs=ctx.env,
-        args=ctx.command,
-        privileged=True,
-        volume_mounts=mounts,
-    )
+    # If ceph-volume creates osd data directories which won't be persisted
+    # so we use a tmp dir for that.
+    # FIXME: probably we can use /var/lib/ceph/{fsid}?
+    with tempfile.TemporaryDirectory() as tmp_osd_dir:
+        # match /var/lib/ceph/osd/ dir permissions
+        os.chown(tmp_osd_dir, uid, gid)
+        os.chmod(tmp_osd_dir, 0o755)
+        # store newly created osds here
+        mounts[tmp_osd_dir] = '/var/lib/ceph/osd/:z'
```

Comment:
@sebastian-philipp @adk3798

Hey guys, I need some of your help in this problem when creating osds. I've gathered some of the things
that I've tried and listed them to discuss them:

_Different approaches to create osds with cephadm binary_

* **ceph orch daemon add osd**
  This approach doesn't seem to work neither in kcli or the box. There is 
  a problem when trying to deploy the osd since the new osd container tries to execute
  ceph-osd with a data directory, where there aren't any of the files created with
  ceph-volume because the data directory is a tmpfs mount. 
  This happens with loopback devices and with typical devices.

* **ceph orch apply** 
    This command works with filters as far as I know. To make this work it would involve
    adding logic to understand either loopback devices or lv from loopback devices
    but, I don't like this that much because loopback devices are used for the sole purpose
    of debugging.

* **cephadm ceph-volume prepare/create/...**
    First of all when calling cephadm ceph-volume to create osds, the container is missing 
    bootstrap-osd keyring so it cannot create osds. Furthermore, when creating a osd I think 
    that the block device (lv of loopback device) is formatted correctly but, the same as
    `ceph orch daemon add osd` happens where without a data directory with the expected files
    we cannot `cephadm deploy --name osd.X`.
    
    In this part of the code you'll find my proposal. If an osd was created with `cephadm ceph-volume` why
    not just map `/var/lib/ceph/osd/ceph-X`(in container) -> `/var/lib/ceph/{fsid}/osd.X`. To sum up,
    copy the osd data directory to the cephadm host and make it usable.
    
    So to deploy it would look like: 
    `cephadm ceph-volume lvm prepare --data /dev/vg1/lv0 --no-tmpfs --no-systemd`
    and then `cephadm deploy --name osd.0`.

* **making data dir manually**
   This would involve to stop messing around with cephadm binary but I haven't made it work completly.